### PR TITLE
DM-7530: Add the science pipeline logging guide

### PR DESCRIPTION
**Pull request by hchiang2**
Preview: https://developer.lsst.io/v/DM-7530/coding/logging.html

### CODE REVIEW

File: coding/logging.rst [hunk 1]
```diff
@@ -0,0 +1,231 @@
+#######
+Logging
+#######
+
+This page provides guidance to developers for using logging with :lmod:`lsst.log`  in the science pipeline code base.
```
Comment (jonathansick) [inline coding/logging.rst]: Science Pipelines

(capitalize and plural; I should make a style guide 😄  )

File: coding/logging.rst [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+Logging
+#######
+
+This page provides guidance to developers for using logging with :lmod:`lsst.log`  in the science pipeline code base. 
+Developers are encouraged to insert log messages whenever and wherever they might be useful, with appropriate component names and levels.
+
+.. _logger-names:
+
+Logger Names
+============
```
Comment (jonathansick) [inline coding/logging.rst]: Add a blank line below the heading's underline.

File: coding/logging.rst [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+Logger names should generally start with the fully qualified name of the module/file containing the logger, without the leading ``lsst.``.
+If the logger is saved as a variable in a class, it is often appropriate to name the logger after the class.
+Logger names use ``.`` as component separators, not ``::``, even in C++.
+Some example logger names are ``afw.image.MaskedImage``, ``meas.algorithms.starSelector``, etc.
```
Comment (jonathansick) [inline coding/logging.rst]: No need to end the sentence with `, etc.`.

Comment (jonathansick) [inline coding/logging.rst]: Actually, I'd recommend moving the `meas.algorithms.starSelector` example as a sentence right after the first "Logger names should generally..." sentence.

Then the note about class and class-based names would be a second paragraph (and the `afw.image.MaskedImage` would be a second sentence of the second paragraph).

The sentence about separators can be on its own as a third paragraph.

File: coding/logging.rst [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+
```
Comment (jonathansick) [inline coding/logging.rst]: just need one blank line between sections

File: coding/logging.rst [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+The simplest way to log is:
+
+.. code-block:: python
+
+   import lsst.log
+   lsst.log.info('Some information during normal operation')
+   lsst.log.warn('Here is a warning!')
+
+
+The example above logs to the root logger.
```
Comment (jonathansick) [inline coding/logging.rst]: You could say "This example logs [...]"

Separate this sentence and "A better practice..." in two separate paragraphs.

File: coding/logging.rst [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+A better practice is to use a named logger following our :ref:`name convention <logger-names>`, and log to that logger in your code to indicate where the logging messages originate.
+For example,
```
Comment (jonathansick) [inline coding/logging.rst]: colon instead of comma

File: coding/logging.rst [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+.. code-block:: python
+
+   import lsst.log
+   lsst.log.info('Some information during normal operation')
+   lsst.log.warn('Here is a warning!')
+
+
+The example above logs to the root logger.
+A better practice is to use a named logger following our :ref:`name convention <logger-names>`, and log to that logger in your code to indicate where the logging messages originate.
```
Comment (jonathansick) [inline coding/logging.rst]: you could replace everything after the comma (include the comma) with

```
to indicate where the logging messages originate.
```

File: coding/logging.rst [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+For example,
+
+.. code-block:: python
+
+   logger = lsst.log.Log.getLogger("meas.algorithms.starSelector")
+   logger.info("This is information about the star selector algorithm execution. %f", 3.14)
+
+In Python, comma-separated arguments are preferred over the string formatting so it is done only if the log record is to be printed.
```
Comment (jonathansick) [inline coding/logging.rst]: I didn't quite understand this sentence. Are you saying not to use `%` formatting, and that `lsst.log` will do the string formatting for you? I.e., is `so it is done only if the log record is to be printed` an action taken by the developer or by `lsst.log` internally?

File: coding/logging.rst [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+==================
+
+To use :lmod:`lsst.log` in C++, the header file to include is:
+
+.. code-block:: c++
+
+   #include "lsst/log/Log.h"
+
+Typically one of the logging macros should be used.
+The developers can choose either the sprintf style formatting (the ``LOGL_`` family) or the iostream style formatting (the ``LOGLS_`` family) to log to a logger.
```
Comment (jonathansick) [inline coding/logging.rst]: "The developers" → "You"

Comment (jonathansick) [inline coding/logging.rst]: You can drop "the" from before "sprintf style" and "isostream style".

File: coding/logging.rst [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+   LOGL_WARN("meas.algorithms.starSelector.psfCandidate", "Failed to make a psfCandidate")
+
+
+Log Levels
+==========
+
+:lmod:`lsst.log` has six levels; in increasing order of severity the are: ``TRACE`` < ``DEBUG`` < ``INFO`` < ``WARN`` < ``ERROR`` < ``FATAL``.
+The guideline of using the log levels is as follows:
+
+- FATAL: for severe errors that may prevent further execution of the component
```
Comment (jonathansick) [inline coding/logging.rst]: You can add full stop (.) after each list item.

File: coding/logging.rst [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+- ERROR: for errors that may still allow the execution to continue
+- WARN: for conditions that may indicate a problem but that allow continued execution
+- INFO: for information that is of interest during normal execution including production
+- DEBUG: for information that is of interest to developers but not of interest during production
+- TRACE: for detailed information when debugging
+
+For loggers used at DEBUG and TRACE levels, it is often desirable to add further components to the logger name; these would indicate which specific portion of the code or algorithm that the logged information pertains to.
+For example,
```
Comment (jonathansick) [inline coding/logging.rst]: colon

File: coding/logging.rst [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+The idea here is that the author understands the intent of the log message and can simply name it, without worrying about its relative importance or priority compared with other log messages in the same component.
+A person debugging the code would typically be looking at it and so would be able to determine the appropriate name to enable.
+The hierarchy allows all components to be easily enabled or disabled together.
+
+
+Logging within the Task framework
+=================================
+
+For Python codes of the pipeline tasks (:lclass:`lsst.pipe.base.Task`), :lattr:`lsst.pipe.base.Task.log` is the logger, so logging looks like this inside subclasses of :lclass:`lsst.pipe.base.Task` or :lclass:`lsst.pipe.base.CmdLineTask`:
```
Comment (jonathansick) [inline coding/logging.rst]: Maybe cleaner to say

```
Pipeline tasks (subclasses of :lclass:`lsst.pipe.base.Task` or :lclass:`lsst.pipe.base.CmdLineTask`) should use the :lattr:`lsst.pipe.base.Task.log` attribute logger:
```

File: coding/logging.rst [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+   DEBUG TRACE3.example.component: Fine tracing to TRACE3
+   DEBUG TRACE4.example.component: Fine tracing to TRACE4
+   DEBUG TRACE5.example.component: Fine tracing to TRACE5
+
+
+Porting from pex_logging
+========================
+
+Logging in the science pipeline codes is being migrated from using :lmod:`lsst.pex.logging` to :lmod:`lsst.log`.
+Below some examples are listed and may be used as a starting point of porting:
```
Comment (jonathansick) [inline coding/logging.rst]: - `Below,`
- `and may be used` → `that may be used`
- `of porting` → `for porting`

File: coding/logging.rst (hunk 1 above)
Comment (timj) [inline coding/logging.rst]: If only the Science Pipelines had a proper name...

File: coding/logging.rst (hunk 8 above)
Comment (hchiang2) [inline coding/logging.rst]: >  Are you saying not to use % formatting, and that lsst.log will do the string formatting for you? 
Yes. ``lsst.log`` does the string ``%`` formatting internally for those messages logged to the level above the set threshold.

Comment (timj) [inline coding/logging.rst]: I do wonder how we can support `format()` syntax so we can be equal opportunity string formatters.

File: coding/logging.rst (hunk 1 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected, here and another instance later.

File: coding/logging.rst (hunk 2 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected.

File: coding/logging.rst (hunk 3 above)
Comment (hchiang2) [inline coding/logging.rst]: Rephrased and reorganized as suggested.

File: coding/logging.rst (hunk 4 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected, here and many instances later.

File: coding/logging.rst (hunk 5 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected.

File: coding/logging.rst (hunk 7 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected as suggested.

File: coding/logging.rst (hunk 6 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected.

File: coding/logging.rst (hunk 8 above)
Comment (hchiang2) [inline coding/logging.rst]: The sentence is rephrased.

File: coding/logging.rst (hunk 9 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected as suggested.

File: coding/logging.rst (hunk 10 above)
Comment (hchiang2) [inline coding/logging.rst]: Added ``.`` for all items.

File: coding/logging.rst (hunk 12 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected as suggested.

File: coding/logging.rst (hunk 13 above)
Comment (hchiang2) [inline coding/logging.rst]: Corrected as suggested.

**Pull request merged**